scoring/score.py

```diff
@@ -13,21 +24,158 @@ def __init__(self, message: str, *, code: str) -> None:
         self.code = code
 
 
+def join_text(strings: Iterable[str], separator: str) -> str:
+    strings = tuple(strings)
+
+    try:
+        *strings, right = strings
+    except ValueError:
+        return ""
+
+    if not strings:
+        return right
+
+    left = ", ".join(strings[:-1])
+    return f"{left} {separator} {strings[-1]}"
+
+
+def join_and(strings: Iterable[str]) -> str:
+    return join_text(strings, "and")
+
+
+def join_or(strings: Iterable[str]) -> str:
+    return join_text(strings, "or")
+
+
 class Scorer:
     def __init__(self, teams_data, arena_data):
         self._teams_data = teams_data
-        self._arena_data = arena_data
+        self._districts = arena_data['other']['districts']
+
+        for district in self._districts.values():
+            district['pallet_counts'] = collections.Counter(
+                district['pallets'].replace(' ', ''),
+            )
+            district['highest'] = district['highest'].replace(' ', '')
+
+    def score_district_for_zone(self, name: str, district: District, zone: int) -> int:
+        colour = ZONE_COLOURS[zone]
+
+        num_tokens = district['pallet_counts'][colour]
+        score = num_tokens * DISTRICT_SCORE_MAP[name]
+
+        if colour in district['highest']:
+            # Points are doubled for the team owning the highest pallet
+            score *= 2
```

Comment:
Is `district['highest']` restricted to only one zone letter at the point of data entry? Either way, I wonder whether we should check to see if there's only 1 team that is highest before rewarding this? I could see a marshal entering e.g. 'GY' for highest when they are both highest.

Reply:
There _should_ be a validation check for that already. If any of the validation checks fail then computing the scores is abandoned.

Reply:
This is validated in the UI for fast feedback (though by design that's non-blocking) and in the checks in the scorer (specifically https://github.com/srobo/sr2025-comp/pull/1/files#diff-b89188034b113b9776c2ad5f796fc9c0621c4dd01fae56f243ed3ce3b7d46f30R101-R114).

Reply:
Okay, regarding the second half of my comment then - do we want those checks? I can see it as being quite valid for a marshal to enter two letters for two pallets that are both stacked on another pallet (despite the 25 mm higher clause) and I think the score entry being able to treat that case as no highest token probably would be useful?

Reply:
Historically we have viewed the scoring logic as applying the _scoring_ rules only, not making judgements about the state of the arena. The checks are intended to catch cases which are incompatible with the rules, rather than to provide an opinion about the state.

From a technical standpoint it's feasible to change that, though I'm not sure it's a good idea. I can see the (IMO small) benefit of recording further information about the state of the arena though.

I think my concern around changing that is that it feels like it could be really easy for the wrong thing to be noted down if there's any lack of clarity around where the judgement-calls will be being applied. I don't have a good feel for how likely that is.

Reply:
Decision - we only want marshals to list a highest token if a *single* token is highest, so only having 1 pallet allowed to be highest is desired